split.go

```diff
@@ -0,0 +1,122 @@
+	if success {
+		val, err = strconv.Atoi(strval)
+	}
+	if !success || err != nil {
+		val = def
+	}
+	return val
+}
+
+// maxHostLen returns the maximum possible length of a server message
+// prefix as defined by the following ABNF in RFC 2812:
+//
+//   [ ":" prefix SPACE ]
+//
+func maxPrefixLen(client *Client) int {
+	// Default values taken from https://modern.ircdocs.horse/
+	nicklen := getIntOption(client, "NICKLEN", 10)
+	userlen := getIntOption(client, "USERLEN", 18)
+	hostlen := getIntOption(client, "HOSTLEN", 63)
+
+	// From RFC 2812:
+	//
+	//   prefix = servername / ( nickname [ [ "!" user ] "@" host ] )
+	//
+	// The code here assumes the `servername` is never used as a
+	// prefix. As this function is concerned with messages send by
+	// ones own client. In accordance with the prefix ABNF from
+	// above, the maximum length is therefore calculated as:
+	//
+	//   ":" <nickname> "!" <user> "@" <host> " "
+	//
+	return 1 + nicklen + 1 + userlen + 1 + hostlen + 1
+}
+
+func splitPRIVMSG(event *Event, maxLen int) (events []*Event) {
+	newMsg := func(text []byte) *Event {
+		e := event.Copy()
+		e.Params[len(event.Params)-1] = string(text)
+		return e
+	}
+
+	// Event used to calculate base length of command, this does not
+	// include the event source and the last parameter.
+	rawEvent := event.Copy()
+	rawEvent.Params = event.Params[0 : len(event.Params)-1]
+
+	// maxTextLen must not be exceeded by the last parameter of the
+	// PRIVMSG. Also include " :" for formatting this last parameter.
+	maxTextLen := maxLen - rawEvent.Len() - len(" :")
+	if maxTextLen <= 0 {
+		return []*Event{event} // TODO: too long without any text
+	}
+
+	b := []byte(event.Last())
+	for len(b) > maxTextLen {
+		idx := bytes.LastIndexByte(b[:maxTextLen], byte(' '))
+		if idx > 0 {
+			// maxTextLen is inclusive → include seperator
+			idx++
+		} else {
+			// maxTextLen is inclusive → include extra byte
+			idx = bytes.LastIndexFunc(b[:maxTextLen+1], utf8.ValidRune)
+		}
+
+		events = append(events, newMsg(b[:idx]))
+		b = b[idx:]
+	}
+	events = append(events, newMsg(b))
+
+	return events
+}
+
+// splitEvent splits a given event into multiple events to satisfy the
+// maximum message length requirements imposed upon the given client by
+// the associated IRC server.
+func splitEvent(client *Client, event *Event) []*Event {
+	// From RFC 2812:
+	//   IRC messages are always lines of characters terminated with a CR-LF
+	//   (Carriage Return - Line Feed) pair, and these messages SHALL NOT
+	//   exceed 512 characters in length, counting all characters including
+	//   the trailing CR-LF.
+	const maxIRClen int = 512 - len("\r\n")
+
+	// We cannot calculate the length of the source part actually
+	// used by the server. Assume the largest possible value and
+	// make sure source is unset for the event to ensure it is not
+	// take into account by `event.Len()`.
+	event.Source = nil // XXX: Operate on a copy instead?
+
+	maxLen := maxIRClen - maxPrefixLen(client)
```

Comment:
The `maxPrefixLen(client)` value should actually remain unchanged for a given connection. This could theoretically be moved to a unexported member field of the client struct.